LVGL: Add native_posix input driver

pavlohamov opened the pull request:
1) Implement native_posix input driver (kscan)
2) Support input in lvgl display sample

Commit:
drivers: kscan: Introduce SDL mouse driver

Introduces a new SDL mouse driver for the keyboard scan (kscan)
interface. Driver is implemented as SDL event filter

Signed-off-by: Pavlo Hamov <pavlo_hamov@jabil.com>

vanwinkeljan commented on drivers/kscan/Kconfig.sdl:
```diff
@@ -0,0 +1,9 @@
+config KSCAN_SDL
+	bool "SDL event filter for touch panel emulation"
+	depends on SDL_DISPLAY
+	default y if SDL_DISPLAY
```
I would not enable this by default even if `SDL_DISPLAY` is enabled

pavlohamov replied:
got it. Done

vanwinkeljan commented on drivers/kscan/Kconfig.sdl:
```diff
@@ -0,0 +1,9 @@
+
+config KSCAN_SDL
+	bool "SDL event filter for touch panel emulation"
+	depends on SDL_DISPLAY
```
Maybe better depend on SDL (`HAS_SDL`) instead of `SDL_DISPLAY`

pavlohamov replied:
Seems to that both will do the same trick. But theoretically SDL could be enabled without SDL_DISPLAY. Done

vanwinkeljan reviewed (changes requested):
Some minor remarks and I also have a feeling the dts config is a bit redundant in this case

pavlohamov commented:
> 
> 
> Some minor remarks and I also have a feeling the dts config is a bit redundant in this case

Don't know why, but i thought it not gonna work without dts entry. Removed. Now style is consistend with SDL display.

jfischer-phytec-iot commented on samples/display/lvgl/boards/native_posix.conf:
```diff
@@ -2,3 +2,8 @@ CONFIG_SDL_DISPLAY=y
 CONFIG_SDL_DISPLAY_X_RES=320
 CONFIG_SDL_DISPLAY_Y_RES=240
+
+CONFIG_KSCAN=y
```
Please move it to `boards/posix/native_posix/Kconfig.defconfig`, it already has DISPLAY configuration there. The same for `native_posix_64.conf`

aescolar replied:
@jfischer-phytec-iot You mean CONFIG_KSCAN_SDL if CONFIG_KSCAN?

jfischer-phytec-iot replied:
@aescolar Yes similar as it was done for mimxrt boards https://github.com/zephyrproject-rtos/zephyr/pull/22055/commits/7ccea0841a3cdb9ede89daa776916b9f7473c9f8

pavlohamov replied:
@jfischer-phytec-iot Does it make sense to add
```
config DISPlAY
	default y if LVGL

config KSCAN
	default y if LVGL
```
as well?

jfischer-phytec-iot replied:
I still do not agree to add new options to samples/*...*/boards/*.conf files. This way It has to be done for every sample that would use display and an input device. And on this architecture is will be always LVGL and CONFIG_LVGL_POINTER_KSCAN, there is nothing else that could be used with it.

vanwinkeljan replied:
@jfischer-phytec-iot 

> I still do not agree to add new options to samples/.../boards/*.conf files

Actually the real problem here is that the basic LVGL sample has been reused to act as touchscreen sample, the best solution in this case is to split both samples and put `KSCAN` in the `prj.conf`.

> And on this architecture is will be always LVGL and CONFIG_LVGL_POINTER_KSCAN, there is nothing else that could be used with it.

I think `native_posix` should enable as least as possible features, e.g. if I would like to test LVGL without touchscreen support I don't want to be forced to disable KSCAN in my project config files.

jfischer-phytec-iot replied:
This sample is to simple to have two version, I suggest to move `CONFIG_KSCAN=y` to prj_kscan.conf

pavlohamov committed:
samples: display: lvgl use posix input, remove defaults

Add support of SDL mouse events as input for the LVGL
platforms: native_posix(_64)

Signed-off-by: Pavlo Hamov <pasha.gamov@gmail.com>

aescolar commented on boards/posix/native_posix/Kconfig.defconfig:
```diff
@@ -15,6 +15,12 @@ config BOARD
+
+config KSCAN
+	default y if LVGL
+
```
This seems a bit wrong to me.
Normally, here in the board defconfig you set default options for the board. So if, for example, a type of peripheral is enabled (say SERIAL), you enable that board default serial driver (UART_NATIVE_POSIX), so the user has a reasonable default configuration when changing target.

But what you are doing here is enabling a peripheral if a library is enabled. This does not fit in the board configuration. If a library needs a peripheral, then it would be a matter of setting in that library kconfig file a `depends on`, or in the worst case a `select`. (So LVGL could have a `depends on DISPLAY` , and/or DISPLAY could have a `default y if LVGL` in drivers/display/Kconfig)
The library kconfig dependencies should not change depending on the target board.

(On the other hand
```
if KSCAN
config KSCAN_SDL
	default y
```
Is the normal thing to do
)

vanwinkeljan replied:
Agree here and LVGL should never be selected by default if DISPLAY and/or KSCAN is enabled.

pavlohamov replied:
This more like a question. I wasn't sure would it be valid to do in such way. Thank you for explanation. Removed

pavlohamov replied:
@jfischer-phytec-iot could you be so kind to clarify options on "LVGL and CONFIG_LVGL_POINTER_KSCAN". For me both points are making sense:
1) Enable by default (but some case may need no Display, or display without Kscan, etc)
2) Enable in each application (copy-paste)

vanwinkeljan commented on boards/posix/native_posix/Kconfig.defconfig:
```diff
@@ -81,6 +87,16 @@ config SDL_DISPLAY
+	default y
+
+config LVGL_POINTER_KSCAN_DEV_NAME
+	default "SDL kscan" if LVGL
```
@aescolar I think this is acceptable and is needed due to the fact that there is no generic way of binding devices to libraries in zephyr, there are some PRs with proposals but no final solution yet.

Another approach is to set the default name of the SDL KSCAN device to `KSCAN`, this is the approach taken by the SDL display.

jfischer-phytec-iot replied:
```
config KSCAN_SDL
	default y
        depends on KSCAN
```
see #22666

pavlohamov replied:
@aescolar Thank you for pointing this. Indeed it is not required. Removed
@jfischer-phytec-iot  "depends on" is much more convenient than "if". Added

vanwinkeljan commented on drivers/kscan/kscan_sdl.c:
```diff
@@ -0,0 +1,108 @@
+
+static struct sdl_data sdl_data;
+
+DEVICE_AND_API_INIT(sdl, CONFIG_LVGL_POINTER_KSCAN_DEV_NAME, sdl_init,
```
Do not use `CONFIG_LVGL_POINTER_KSCAN_DEV_NAME` as device name as you can not assume that LVGL is enabled.

Better introduce a new Kconfig to set the device name.

pavlohamov replied:
Make sense. LVGL_POINTER_KSCAN_DEV_NAME -> SDL_POINTER_KSCAN_DEV_NAME default LVGL_POINTER_KSCAN_DEV_NAME if LVGL

aescolar reviewed (approved):
+1 Thanks. I see no issue with this, but I leave the +2s for @vanwinkeljan or others more involved in the display area

pavlohamov commented:
rebased on master

pavlohamov commented:
@franciscomunoz @albertofloyd @scottwcpg any comments?

aescolar commented:
@pavlohamov Note that the merge window is closed now freeze until the 2.2 release. So by default this PR will be merged when it reopens again

franciscomunoz commented:
@pavlohamov This API was intended for laptop keyboards. I see you are doing a mouse.

https://docs.zephyrproject.org/latest/reference/peripherals/kscan.html

This PR doesn't  play very well with the intend of the documentation. But I don't have the last word.

pavlohamov commented:
> @pavlohamov This API was intended for laptop keyboards. I see you are doing a mouse.
> 
> https://docs.zephyrproject.org/latest/reference/peripherals/kscan.html
> 
> This PR doesn't play very well with the intend of the documentation. But I don't have the last word.

Thank you for noticing. I was referencing to https://github.com/zephyrproject-rtos/zephyr/pull/22055/  
Unfortunate there no "touch" API right now

MaureenHelm commented:
> @pavlohamov Note that the merge window is closed now until the 2.2 release. So by default this PR will be merged when it reopens again

The merge window is back open. Will you please rebase to retrigger CI checks?

pavlohamov commented:
> > @pavlohamov Note that the merge window is closed now until the 2.2 release. So by default this PR will be merged when it reopens again
> 
> The merge window is back open. Will you please rebase to retrigger CI checks?

Thank you for letting me know. Done
